WIP: A more sophisticated set of blueprints

mphasize opened the pull request:
Hej, I've been working on an "advanced" set of blueprints for some time now and I think it's time to discuss this.

As a preparation I enabled the generator to support different "flavors" of blueprints and I'm keeping the original blueprints around as a "basic" flavor and added the new blueprints as an "advanced" flavor.

So what do these new blueprints do?
- enable full control on how the API REST handles associations/relationships: **url links**, **indexing** and **sideloading** supported.
- add `meta` information to `find` queries, which enable **pagination** on the client side

What needs to be done?
- [x] remove the need to install `lodash` and `pluralize` as separate dependencies
- [x] update blueprints with latest fixes from Sails v0.11
- [x] add the camelCase fix for sideloaded association names in the JSON output
- [x] gather some feedback from the community: does this work/feel right? maybe some better naming for configuration options?
- [x] test how this works on Mongo/NoSQL databases
- [x] optimized performance of the **index** option on Many-to-Many relationships, i.e. on SQL we can just load the indexes from the join table without actually performing a full join.
- [x] simplify `buildResponse` to remove single record output in object style. Instead use array, as it is apparently preferred by Ember Data even for single records.

Happy to get some feedback on this and do some further planning...

mphasize committed:
Support multiple „flavors“ of blueprints

mphasize committed:
Some comments for the generator file

mphasize committed:
Adding a flavor of „advanced“ blueprints

mphasize committed:
Some initial information on the advanced blueprints.

mphasize commented:
@Globegitter please take a look at this and let me know what you think? (Also... I need to add the camelCase fix here...)

wayne-o commented:
👍

Globegitter commented:
@mphasize Good that you are starting a discussion here. After seeing https://github.com/json-api/json-api/pull/341 and https://github.com/emberjs/data/issues/1988#issuecomment-74145489 I was thinking maybe we can have a set of jsonapi blueprints and also rename it to that, to leave it more open to a wider audience that wants to use jsonapi.

I haven't looked at the latest specs properly yet and it is of course still WIP but I think it would be the most future-proof decision.

So what are your thoughts about that? Also /cc @wayne-o @kriswill @mgenev 

But back to your original proposal, how exactly would the differentiation be between basic and advanced? Would that be two different `sails generate` commands? I think that is a great idea - what do you think about embedded support?

Will also take a look at the code itself now :)

mgenev commented:
I like this idea a lot and I will gladly work on it. I was extending the current blueprints yesterday to support geo querying with mongo, postgres has that ability too with with postGIS.

I know that @kriswill has a good convention for a pod-like api structure which would be awesome.

kriswill commented:
For some reference on [JSON:API](http://jsonapi.org) take a look at the [Endpoints](https://github.com/endpoints/endpoints) project and it's [Example](https://github.com/endpoints/example) project.  This implementation of the spec is still young, but is the most robust one so far.

mphasize commented:
Wonderful! So in order:
- JSON API: when I originally started implementing these blueprints I wanted to adapt the json-api standard, but soon had to realise that Ember was not (and still is not) a 100% compliant. My goal was to support Ember's DS.RESTAdapter out-of-the-box as much as possible, so these blueprints turned out quite Ember specific. My idea would be to evolve the blueprints along with the development of Ember Data, getting closer to the (eventually final?) json-api standard step by step.
- Flavors: the differentiation between "basic" and "advanced" is essentially the amount of setup you need to do. For the sake of not breaking stuff, I didn't touch the basic blueprints at all, but I think that eventually (after double-checking) the advanced blueprints can be set to a default configuration that emulates the basic behaviour, so that an "upgrade" is possible without breaking existing code.  
  The original idea of flavors goes a bit further though, because I would really love to support the "embedded" records api style that Ember can consume (with some setup) and I would also love to support a "pure" json-api flavor. If that warrants renaming the whole project or branching a separate generator for sails – no idea.  
  On the practical side: at the moment it's simply another command line argument to pass to the generator.
- Geo querying: nice! I've worked on some geovis projects in the past and I can get easily excited for this kind of stuff. @mgenev Please share what you got! :-)

kriswill commented:
Also check out [ember-data-endpoints](https://github.com/endpoints/ember-data-endpoints) which is being done by @bmac, core ember and ember-data contributor, and the original author of the [ember-data-sails-adapter](https://github.com/bmac/ember-data-sails-adapter) project.

wayne-o commented:
I am really interested in this thread. It all sounds great. I am going to
be busy with my son the next couple so wanted to chime in and make my
interest explicit! I'll be very happy to contribute towards this after the
weekend :)

On Thu, Feb 19, 2015 at 7:56 PM, Kris Williams notifications@github.com
wrote:

> Also check out ember-data-endpoints
> https://github.com/endpoints/ember-data-endpoints which is being done
> by @bmac https://github.com/bmac, core ember and ember-data
> contributor, and the original author of the ember-data-sails-adapter
> https://github.com/bmac/ember-data-sails-adapter project.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/mphasize/sails-generate-ember-blueprints/pull/9#issuecomment-75123963
> .

## 

_--------------------_

_w://_

_t: 07508 215 459_

_--------------------_

mphasize commented:
Some more info / questions around configuring json api output:

In the new blueprints I am introducing 2 ways to configure how associations are handled in the json output: 
- in the global sails configuration
- on a model attribute

In `myapp/config/models.js`

``` javascript
module.exports.models = {
  associations: {
    list: "link",
    detail: "record"
  }
};
```

An in `myapp/api/models/user.js`

``` javascript
module.exports = {

  attributes: {
    name : "string",
    posts: {
      collection: "post",
      via: "user",
      includeIn: {
        list: "record",
        detail: "record"
      }
    }
  }
};
```

Depending on the blueprint, it will either use the `list` or `detail` value. The current mapping is

```
create ---> detail
find ---> list
findOne --> detail
update --> detail
```

What's your take on such fine-grained control and specifying this on the `model` ?
What do you think about the naming of `list`, `detail` and `includeIn` ... ?

albertosouza commented:
@mphasize lets make a sails hook for this feature !
something like sails-hook-ember-blueprint
Sails hook has suport to updates with npm modules i can start the structure and setup it

mphasize commented:
@albertosouza I think the scope of the "advanced" blueprints is a bit broader than what I can see in the other hooks you mentioned here https://github.com/mphasize/sails-ember-blueprints/issues/4 
Instead of just parsing and changing the result of the final controller output, these blueprints aim to control json output depending on the actual request use-case (i.e. find a list, vs. get record details) and they also want to accept data from Ember (for create/update/etc.).

Or did I miss something?

albertosouza commented:
@mphasize yes, you are right but we can create a hook to add these blueprints in the sails, i can help with the development of the hook as it will be useful for my projects

change sails-generate-ember-blueprints to sails-hook-ember-blueprints

mphasize committed:
Fix create blueprint to *really* return a 201 status code.

Setting `res.status(201)` but then using `res.ok()` afterwards would reset the status code to 200, which we don’t want.

mphasize committed:
Fix: camelCase usage for multi-word model names in ember output / added this as a function to the Ember service.

mphasize committed:
Generate „shallow“ blueprints, which allows us to use the bundled npm dependencies and removes the necessity to install `lodash` and `pluralize` after installing the generator.

mphasize committed:
Updated README information

mphasize committed:
Remove fs-extra dependency.

mphasize commented:
@Globegitter Any more thoughts on this?

I just added the "shallow" blueprint generation we discussed earlier...

mphasize committed:
Bring over some fixes from Sails 0.11 blueprints

Globegitter commented:
@mphasize Not quite sure about the different benefits of hook vs generators at this point (even though I am using hooks and liking them). It would not fill up the blueprints folder then?

And will look over the naming tomorrow if it makes intuitive sense to me :)

Globegitter commented:
And yes :+1: :+1: on tests :P Could potentially help on that front a bit.

Globegitter commented on templates/advanced/api/services/Ember.js:
```diff
@@ -0,0 +1,145 @@
+		} );
+	},
+
+	emberizeModelIdentity: function ( model ) {
```
Now that we are using lodash 3.x we don't need that helper function anymore and could just use https://lodash.com/docs#camelCase

Globegitter commented:
@mphasize Sorry I just committed what should have been a PR. But it fixes everything that is mentioned in https://github.com/mphasize/sails-generate-ember-blueprints/pull/8 and in the comment above.

Do you want me to revert, or is it fine like that? I do have this in my project currently and is working well.

mphasize commented:
No worries. I will hold the next "publish" and version bump until we got it all figured out and we need the fix anyway. Is the whole camelCase issue resolved with that?

Globegitter commented:
As far as I can tell yes. But also I haven't taken a step back yet and though it these are all the cases that could come up.

mphasize commented:
I think this mostly covers it and there are even some lines of code where we need to avoid the camelCased version, i.e. when generating url style `links` to associated records.

mphasize commented:
Btw. there's One More Thing that I sneaked into the advanced blueprints and I'm not sure if it's the right place, but I couldn't think of any better while developing this...

Sails normally uses policies to grant or reject access to routes, but sometimes you need to look at a record to even determine whether a user has access or not. And in those cases I found Sails policies either lacking, or not really intended for the job (as I would have to retrieve the record twice). So I built a set of additional hooks into the advanced blueprints which call methods on a model's definition, not unlike the lifecycle methods.

There's a `accessControl` function inside the actionUtil.js that checks the model for any defined callback functions and passes the request _criteria_ and some context to the model for further manipulation. 

I.e. you could define a `controlBeforeFind` method in a _post_ model and change the query parameters to make sure that only posts are returned, that the current user is allowed to see.

``` javascript
    controlBeforeFind: function ( criteria, options ) {
        // modify criteria if necessary...
        return criteria;
    },
```

For me this has been working pretty well so far, but I'm not sure if it really is the right place to handle such logic. Anybody knows how others have handled more complex access control schemes in Sails apps?

Globegitter commented:
@mgenev @kriswill What do you think about @mphasize  comment above? I think that sounds pretty good and makes sense to have.

bmac commented on templates/advanced/api/services/Ember.js:
```diff
@@ -0,0 +1,145 @@
+		var documentIdentifier = plural ? modelPlural : emberModelIdentity;
+		var json = {};
+
+		json[ documentIdentifier ] = plural ? [] : {};
```
Not sure if this will simplify your code at all but the Ember Data RESTSerializer allows you can return a single resource wrapped in an array. Returning an object without the array still works but it is considered more of a legacy thing https://github.com/emberjs/data/blob/master/packages/ember-data/lib/serializers/rest_serializer.js#L283-L287

mphasize replied:
Very good to know! Thanks for pointing this out!

Globegitter commented:
I just came up for another need, that I am not sure if/how it is supported currently. Maybe you know more @mphasize I have a model with relationships and just want to return models who actually have that relationship.

So how you can now do `http://localhost:1337/api/v1/project?category=bla` with a normal attribute, I would love to be able to do `http://localhost:1337/api/v1/project?topics=true` or something, where-as topics is a relationship attribute in the model. Not that I expected it to work, but right now it throws `ER_BAD_FIELD_ERROR: Unknown column 'researchproject.funds' in 'where clause'` using the mysql adapter.

Any idea if such a thing would be possible easily?

Globegitter commented:
Also I have just been looking a bit over the namings/settings you mentioned above (not the actual code though) and I do like the possiblity to have one global setting but also being able to set it per model and even per find/update/etc.

I don't fully understand the naming itself right now but that does not necessarily mean that it is unintuitive :)

mphasize commented:
@Globegitter I will take a look and see what I can find out about how this will be handled server side, but I'd be curious how you would even generate such a request through the Ember REST Adapter?

I'll be back on this towards the end of the week, hoping that I can incorporate all the feedback and maybe even finish this PR. Until then my day job is keeping me busy. ;)

Globegitter commented:
@mphasize You can just send arbitrary flags to the server, for example `this.store.find('researchProject', { limit: 50, funds: true });` creates `http://127.0.0.1:1337/api/v1/researchProjects?funds=true&limit=50`

Though filtering associations seems to be harder within waterline, just playing around with it and see where I'll get. See: https://github.com/balderdashy/waterline/issues/855

Also it would be cool to be able to turn sideloading off that way: I am trying here to really optimize the data I am sending to the client, because I am sending a lot of data already, so all that I can reduce really is a win and I actually don't need sideloading for a certain request. This actually seems quite simple by just adding:

``` js
  performSideload = req.params.performSideload || performSideload //not sure if that is actually the right syntax
    res.ok( actionUtil.emberizeJSON( Model, matchingRecords, req.options.associations, performSideload ) );
```

or something slightly more sophisticated.

And imo even more important would be something like `?select[]=fieldname1&select[]=fieldname2` to further optimize the json size. Ok for that see: https://github.com/balderdashy/waterline/pull/509

Also I would love to help some more whenever I find some time with actual code writing :)

Globegitter commented:
It would be nice to also have a separate helperfunction, `toEmberJSON(records)` or `toJSONApi(records), that just take a record or an array of records, returned by`Model.find()` and turns it into an ember json. 

Edit: Ok `emberizeJSON` actually works quite well just seemed a bit more complicated with all the extra parameters ;)

This could also have sideload support, purely by looking at the model and then manipulating the given object but that is not that important imo.

Also just playing around a bit more with the blueprints in my project, so actually getting to know them better.

How do you actually perform the sideload? That is one bit | haven't fully figured out yet.

wayne-o commented:
Hi guys, what state is this in atm? Can i point my code at it and have a play or is there more work needed? Is there anything I can help with?

mphasize commented:
@wayne-o Go for it! There's lots of ideas and discussion (and therefore changes to come), but the current code should be working.

wayne-o commented:
I was already on it lol!

WORKS LIKE A CHARM!!!!

Good job guys - seriously!

On Wed, Feb 25, 2015 at 7:13 PM, Marcus notifications@github.com wrote:

> @wayne-o https://github.com/wayne-o Go for it! There's lots of ideas
> and discussion (and therefore changes to come), but the current code should
> be working.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/mphasize/sails-generate-ember-blueprints/pull/9#issuecomment-76033562
> .

## 

_--------------------_

_w://_

_t: 07508 215 459_

_--------------------_

mphasize commented:
@Globegitter I will need a bit more time to think through all the ideas and suggestions - hopefully towards the end of the week, but maybe I can answer this straight away:

How the sideload is performed:
- in the find or findOne blueprint we check the associations defined on the model
- in the advanced blueprint we furthermore check how these associations should be included
- if an association is supposed to be indexed or embedded/sideloaded, we use an adapted version of the original `populateEach`, telling the Waterline find query to populate the associations (hence a minimum version of Sails 0.10 required)
- after the find query comes back with all the associations embedded (since that is the original Waterline behavior) the `buildResponse` function (formerly known as `emberizeJSON`) gets to work on the data structure, analyses the model again for association declarations and pulls the records out into a sideloaded data structure. (It really is just shuffling objects around in a tree).
- for associations that shall be indexed, the records get flattened into ID arrays (future optimization would be to not even load the records - just the indexes, see [waterline#532](https://github.com/balderdashy/waterline/issues/532))
- for associations that shall only be linked to (we didn't populate the query) and now we generate a `links` object with the urls to load further data

Does this clarify what happens for you?

mphasize commented:
@Globegitter Btw. at the point where we tell Waterline which associations to populate, we could look for further parameters in the API request to change what is included/sideloaded.

mphasize commented:
@wayne-o Thanks and good to hear that it's working. :)

mphasize commented:
Alright... back to this again. 
As we can see with #13 and my own ambitions regarding the `accessControl` hooks into models, there's a need to dive deeper into the relationship between request and model and what's supposed to happen.

However, it feels like we're overloading the scope of the blueprints when we introduce all of this, so I took a step back and had another look at Sails policies (because that's where stuff like access control) should belong according to the Sails philosophy. Digging deeper I found a couple of convenient things that Sails sets in req.options and that is accessible in the policies, so I am pretty confident that I can create a policy that handles the accessControl stuff and links to functions in the Model or wherever we want to put them. It should also allow us to keep the blueprints neat and clean. I'll do some further experiments today and if everything turns out well, I will probably separate the accessControl efforts into their own project.

mphasize committed:
Merge master to bring in the latest updates

mphasize committed:
Remove controlBefore*Blueprint hooks, these are better handled in a policy.
Remove the inject user into create record values (again this fits better into a policy).

mphasize committed:
Added some description and links to the Policy gists…

mphasize committed:
Switch to lodash’s camelCase method

mphasize commented on templates/advanced/api/blueprints/_util/actionUtil.js:
```diff
@@ -0,0 +1,438 @@
+	 * @param  {[type]}   associations [description]
+	 * @param  {Function} done         [description]
+	 */
+	populateIndexes: function ( parentModel, ids, associations, done ) {
```
Can anybody verify that this method runs as expected on a Mongo/NoSQL database? (It's used in `find` and `findOne` blueprints). Set sideloading to `index` to enable this.

Globegitter replied:
@mgenev Maybe you can give that a try at some point?

wayne-o replied:
I should be able to give this a spin later today

On Mon, Mar 2, 2015 at 2:50 PM, Markus Padourek notifications@github.com
wrote:

> In templates/advanced/api/blueprints/_util/actionUtil.js
> https://github.com/mphasize/sails-generate-ember-blueprints/pull/9#discussion_r25602915
> :
> 
> > -           console.log( "Dev: Model " + model.identity + " is missing control " + action + "!" );
> > -           if ( callback ) {
> > -               callback( null, primaryArgument );
> > -           }
> > -       }
> > -       return primaryArgument;
> > -   },
> >   +
> > -   /**
> > -    \* helper function to populate a record with an array for indexes for associated models, running various Waterline queries on the join tables if neccessary ( defined as: include -> index )
> > -    \* @param  {Waterine Collection}   parentModel  [description]
> > -    \* @param  {Array|Integer}   ids          [description]
> > -    \* @param  {[type]}   associations [description]
> > -    \* @param  {Function} done         [description]
> > -    */
> > -   populateIndexes: function ( parentModel, ids, associations, done ) {
> 
> @mgenev https://github.com/mgenev Maybe you can give that a try at some
> point?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/mphasize/sails-generate-ember-blueprints/pull/9/files#r25602915
> .

## 

_--------------------_

_w://_

_t: 07508 215 459_

_--------------------_

mphasize committed:
Add `linkAssociations` for sideloaded records (2nd level associations)

mphasize commented on templates/advanced/api/services/Ember.js:
```diff
@@ -0,0 +1,143 @@
+					if(!_.isNumber(array[0]) && !_.isString(array[0])) { // this is probably an array of records
+						var model = sails.models[ pluralize(key, 1) ];
+						console.log("Sideloaded records model: ", model.identity);
+						Ember.linkAssociations(model, array);
```
This should fix #7

mphasize commented:
Alright, this has been an enlightening discussion throughout for me and inspired a lot of good stuff, but this PR is getting a bit heavy and difficult to navigate, so...

I documented further todos in the roadmap issues #17 and #18. Please feel free to add more suggestions and continue discussion over there. Since this PR will not change existing setups, I'm going to merge it into master as soon as @mgenev or @wayne-o give us (hopefully) positive feedback on MongoDB. Then we can tackle the remaining issues in other PRs.

Globegitter commented:
@mphasize Yep that sounds like a good way forward :)

mphasize committed:
Fixed a bug where sideloaded records would be sideloaded but also fully included in the primary record.

mphasize commented:
Any news on Mongo?

mgenev commented:
Sorry, no, I set everything up for a test on monday, but had to go to emberconf, now I'm back and I should be able to do the test finally.

mphasize committed:
Looks like it is not a good idea to create `links` for belongsTo (hasOne) associations

Also, we can always use the plural identifier for records

mphasize committed:
Prune empty association arrays in JSON output

mgenev commented:
@mphasize i'm not seeing if i'm installing from the branch correctly. I downloaded this branch and put it in my new project's node_modules dir, I did generate advanced, and the crud methods now point to the /node_modules/sails-generate-ember-blueprints. However, the /api/blueprints/_util/actionUtil remained the old one and when i try to lift I get:

```
module.js:322
    throw err;
    ^
Error: Cannot find module 'lodash/array/uniq'
    at Function.Module._resolveFilename (module.js:320:15)
    at Function.Module._load (module.js:262:25)
    at Module.require (module.js:349:17)
    at require (module.js:368:17)
    at Object.<anonymous> (/Users/martin.genev/Projects/advanced-sails-ember-blueprint/api/blueprints/_util/actionUtil.js:5:21)
    at Module._compile (module.js:444:26)
    at Object.Module._extensions..js (module.js:462:10)
    at Module.load (module.js:339:32)
    at Function.Module._load (module.js:294:12)
    at Module.require (module.js:349:17)
    at require (module.js:368:17)
    at /Users/martin.genev/Projects/advanced-sails-ember-blueprint/node_modules/include-all/index.js:129:29
    at Array.forEach (native)
    at requireAll (/Users/martin.genev/Projects/advanced-sails-ember-blueprint/node_modules/include-all/index.js:44:9)
    at /Users/martin.genev/Projects/advanced-sails-ember-blueprint/node_modules/include-all/index.js:54:23
    at Array.forEach (native)
```

am i missing a step somewhere, doing something wrong?

**_EDIT**_
I copied the new action util manually and ran install lodash and util and got it to work.

mgenev commented:
@mphasize If I understood you correctly you wanted to see if the array of index as an option works with mongo. To do this, I set this in the global model config:

```
associations: {
    list: "index",
    detail: "record"
  },
```

This did indeed with no visible error produce the array like this (vendors belong to users):

```
{  
   users:[  
      {  
         password:"$2a$10$2j4HYXFSJY3HIwlQVaISBuyjoT.Dn/uWp3lcYrr..VxJ/dmhnsX46",
         username:"email@gmail.com",
         firstName:"martin",
         lastName:"genev",
         userType:0,
         createdAt:"2015-02-23T03:14:06.941Z",
         updatedAt:"2015-02-23T03:14:06.941Z",
         id:"54ea9afe0ac0c5c35ca30244",
         vendors:[  
            "54ec0029ca7f8a94b15182d7",
            "54efc5bfc7db589a6e521af1",
            "54efc5edc7db589a6e521af2"
         ]
      },
      {  
         password:"$2a$10$pToIzzVEMI.SRBlRJb2SCODqLLWwbddVn1dkVe/nDN/udkkN7gDMG",
         username:"anotheremail@gmail.com",
         createdAt:"2015-02-27T05:02:51.636Z",
         updatedAt:"2015-02-27T05:02:51.636Z",
         id:"54effa7b4fc3b5892da2f041",
         vendors:[  

         ]
      }
   ],
   meta:{  
      total:2
   }
}
```

After that I switched to:

```
associations: {
    list: "record",
    detail: "record"
  },
```

And this once again with no error produced the entire records sideloaded:

```
{  
   users:[  
      {  
         vendors:[  
            "54ec0029ca7f8a94b15182d7",
            "54efc5bfc7db589a6e521af1",
            "54efc5edc7db589a6e521af2"
         ],
         password:"$2a$10$2j4HYXFSJY3HIwlQVaISBuyjoT.Dn/uWp3lcYrr..VxJ/dmhnsX46",
         username:"email@gmail.com",
         firstName:"martin",
         lastName:"genev",
         userType:0,
         createdAt:"2015-02-23T03:14:06.941Z",
         updatedAt:"2015-02-23T03:14:06.941Z",
         id:"54ea9afe0ac0c5c35ca30244"
      },
      {  
         vendors:[ ],
         password:"$2a$10$pToIzzVEMI.SRBlRJb2SCODqLLWwbddVn1dkVe/nDN/udkkN7gDMG",
         username:"another@gmail.com",
         createdAt:"2015-02-27T05:02:51.636Z",
         updatedAt:"2015-02-27T05:02:51.636Z",
         id:"54effa7b4fc3b5892da2f041"
      }
   ],
   vendors:[  
      {  
         name:"mhm",
         description:"981118314",
         urlSegment:null,
         address:"Seattle, WA",
         location:{  
            type:"Point",
            coordinates:[  
               -122.3320708,
               47.6062095
            ]
         },
         user:"54ea9afe0ac0c5c35ca30244",
         createdAt:"2015-02-24T04:38:01.790Z",
         updatedAt:"2015-02-24T04:38:01.790Z",
         id:"54ec0029ca7f8a94b15182d7"
      },
    ...
    /// rest of vendors records trunc'd for paste ///
   ],
   meta:{  
      total:2
   }
}

```

As you can see above I'm also getting meta total. I tested 'link' as well and it produced:

```
links: {
    vendors: "/api/v1/users/54ea9afe0ac0c5c35ca30244/vendors"
}
```

which I then hit in the browser and it gave me the list of the vendors for that user with no hesitation.

Applause! This is great work and I'll be including it in app as soon as you merge it and give it a thorough test over time in different situations.

My question is, and excuse me if the docs answer it but i haven't noticed, do I have a way of altering this behavior from Ember, on a per request basis like this:

```
this.find('users', {assocations: {list: 'index'}}
```

Or whether this clicks with the ember data's 'async' flag on associations I set on the client. This way I could finely control my app's loading behaviors and load whole records async later if I need to as each situation calls for.

Once again, this is awesome!

mphasize commented:
@mgenev Wonderful news! Thanks a lot for trying it out. The results look the way they should be, so I assume that there's at least no general problem and we will have to iron out bugs when they manifest. 

Concerning your question: At the moment you cannot control which associations will be included on a per-request basis. It's on the roadmap, but it's not in the blueprints yet. There's a bit more security thinking necessary, as people requesting data from your API can easily change the query and potentially create a situation with a lot of heavy includes.
However, at the moment you can control behaviour at a per-property basis on each model individually using the `includeIn` keyword.

Example:

``` javascript
// in /api/models/user.js
module.exports = {
    attributes: {
                // ...
        posts: {
            collection: "post",
            via: "user",
            includeIn: {
                list: "link",
                detail: "link"
            }
        }
    }
}
```

Thanks again for your feedback! I'll go ahead and merge this into master now, then I'll give it a bit more time to stew before I release it to npm.

mgenev commented:
Excellent. Please give me heads up when it's on npm and I'll add it to https://github.com/mgenev/how-to-sane which would allow a whole bunch of people to report bugs. I'll be away from computers for 2 weeks starting tonight, but i'll try to keep up with reading on my tablet
